python/triqs_cthyb/solver.py

```diff
@@ -145,11 +148,39 @@ def solve(self, **params_kw):
         # (only supported for G_tau, to permit compatibility with dft_tools)
         if perform_post_proc and (self.last_solve_parameters["measure_G_tau"] == True):
 
+            if self.last_solve_parameters["measure_density_matrix"]:
+                # we have the density matrix, so we will compute the high frequency
+                # moments and orbital occupations
+
+                self.orbital_occupations = orbital_occupations(self.density_matrix,
+                                                               self.gf_struct,
+                                                               self.h_loc_diagonalization
+                                                               )
+
+                h_int = self.last_solve_parameters['h_int']
+                self.Sigma_moments = sigma_high_frequency_moments(self.density_matrix,
+                                                 self.h_loc_diagonalization,
+                                                 self.gf_struct,
+                                                 h_int
+                                                 )
+
+                self.Sigma_Hartree = {bl: sigma_bl[0] for bl, sigma_bl in self.Sigma_moments.items()}
+
+                self.G_moments = green_high_frequency_moments(self.density_matrix,
+                                                         self.h_loc_diagonalization,
+                                                         self.gf_struct,
+                                                         self.h_loc
+                                                         )
+
             # Fourier transform G_tau to obtain G_iw
             for bl, g in self.G_tau:
                 bl_size = g.target_shape[0]
                 known_moments = make_zero_tail(g, 4)
-                known_moments[1,...] = np.eye(bl_size)
+                if self.last_solve_parameters["measure_density_matrix"]:
+                    known_moments[1,...] = self.G_moments[bl][0]
+                    known_moments[2,...] = self.G_moments[bl][1]
+                else:
+                    known_moments[1,...] = np.eye(bl_size)
```

Comment:
After changing the convention for the moments we can simplify
```suggestion
                if self.last_solve_parameters["measure_density_matrix"]:
                    known_moments = self.G_moments[bl]
                else:
                    known_moments = make_zero_tail(g, 4)
                    known_moments[1] = np.eye(bl_size)
```
and delete line `known_moments = make_zero_tail(g, 4)` above